net/sqlwriter.c

```diff
@@ -209,13 +206,72 @@ static int from_timeout_cb(struct sqlwriter *writer)
 
 static int sql_pack_response(struct sqlwriter *writer, int total, void *arg)
 {
-    struct iovec v[1];
-    if (evbuffer_reserve_space(writer->wr_buf, total, v, 1) == -1) {
-        return -1;
+    int rc = writer->pack(writer, arg);
+    /*
+     * rc > 0 : done
+     *    = 0 : not done yet
+     *    < 0 : error
+     */
+    if (rc >= 0) {
+        writer->done = (rc > 0);
+        rc = 0;
     }
-    v[0].iov_len = total;
-    writer->done = writer->pack(v[0].iov_base, arg);
-    return evbuffer_commit_space(writer->wr_buf, v, 1);
+    return rc;
+}
+
+/*
+ * An 'append' callback to consume 'packed' data as it is being generated.
+ * The function flushes every `max_buf' many bytes, hence keeps the memory use
+ * of the writer's evbuffer under `max_buf'.
+ *
+ * This is how the function gets invoked:
+ * sql_write -> sql_pack_response -> newsql_pack -> sql_append_packed
+ */
+int sql_append_packed(struct sqlwriter *writer, const void *data, size_t len)
+{
+    /* NOTE: we're holding writer->wr_lock */
+
+    size_t nleft = len;
+    struct evbuffer *wr_buf = writer->wr_buf;
+    const uint8_t *ptr = data;
+    int rc;
+
+    while (nleft > 0) {
+        if (evbuffer_get_length(wr_buf) >= max_buf) {
+            /* We've accumulated enough bytes, flush now. */
+            Pthread_mutex_unlock(&writer->wr_lock);
```

Comment:
Can heartbeat event add a h/b packet in the middle of half written response at this point?

Reply:
I guess simple test would be to sleep 5 sec after unlock to ensure trickle drains and enqueues a h/b packet.

Reply:
Yes I overlooked it. Heartbeat can definitely sneak in during the period where we give up and re-acquire wr_lock. I'll change this.